sql: make the schema changer drop ADDING tables in rollbacks

rohany opened the pull request:
Fixes #52828.

This commit teaches the schema changer to clean up tables in the
`ADDING` state during rollbacks.

Release note (bug fix): Fix a bug where tables would not be cleaned up
after a failed creation through `CREATE TABLE AS`.

rohany committed:
sql: make the schema changer drop ADDING tables in rollbacks

Fixes #52828.

This commit teaches the schema changer to clean up tables in the
`ADDING` state during rollbacks.

Release note (bug fix): Fix a bug where tables would not be cleaned up
after a failed creation through `CREATE TABLE AS`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52832)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52832#-:-MEwsM782gGomN4lqfa6:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52832#-MEwrfuU4tOmdtbNzmta:-MEwrfuV5BHIM444dbgx:b7zil79) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea7ff14ed5c00c2903eb4239e7b0893c71969309/pkg/sql/schema_changer.go#L414)):*
> ```Go
> 
> 			// Reclaim all old names.
> 			for _, drain := range namesToReclaim {
> ```

The fact that we don't batch these is likely a regression, can you add a `TODO(ajwerner)` to batch these?
___
*[pkg/sql/schema_changer_test.go, line 463 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52832#-MEws0_c1u-YorAAtmw1:-MEws0_c1u-YorAAtmw2:bxueo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea7ff14ed5c00c2903eb4239e7b0893c71969309/pkg/sql/schema_changer_test.go#L463)):*
> ```Go
> 	params, _ := tests.CreateTestServerParams()
> 	params.Knobs = base.TestingKnobs{
> 		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
> ```

slight worry that this may end up being flakey because it gets called on the wrong job when a new job is introduced into the system. Seems like a sort of bad API to just pass the job id but it is what it is. I'd be great to verify that this is the job you created. 


<!-- Sent from Reviewable.io -->

rohany commented:
> The fact that we don't batch these is likely a regression, can you add a TODO(ajwerner) to batch these?

done

> slight worry that this may end up being flakey because it gets called on the wrong job when a new job is introduced into the system. Seems like a sort of bad API to just pass the job id but it is what it is. I'd be great to verify that this is the job you created.

Done, added a different testing knob and switched to using it.

rohany commented:
bors r+
